port-forwarder/fwd_tcp.go

```diff
@@ -0,0 +1,230 @@
+package port_forwarder
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/slackhq/nebula/service"
+	"golang.org/x/sync/errgroup"
+)
+
+type ForwardConfigOutgoingTcp struct {
+	ForwardConfigOutgoing
+}
+
+func (cfg ForwardConfigOutgoingTcp) ConfigDescriptor() string {
+	return fmt.Sprintf("outbound.tcp.%s.%s", cfg.localListen, cfg.remoteConnect)
+}
+
+type ForwardConfigIncomingTcp struct {
+	ForwardConfigIncoming
+}
+
+func (cfg ForwardConfigIncomingTcp) ConfigDescriptor() string {
+	return fmt.Sprintf("inbound.tcp.%d.%s", cfg.port, cfg.forwardLocalAddress)
+}
+
+type PortForwardingOutgoingTcp struct {
+	l                     *logrus.Logger
+	tunService            *service.Service
+	cfg                   ForwardConfigOutgoingTcp
+	localListenConnection *net.TCPListener
+}
+
+func (fwd PortForwardingOutgoingTcp) Close() error {
+	fwd.localListenConnection.Close()
+	return nil
+}
+
+func (cf ForwardConfigOutgoingTcp) SetupPortForwarding(
+	tunService *service.Service,
+	l *logrus.Logger,
+) (io.Closer, error) {
+	localTcpListenAddr, err := net.ResolveTCPAddr("tcp", cf.localListen)
+	if err != nil {
+		return nil, err
+	}
+	remoteTcpAddr, err := net.ResolveTCPAddr("tcp", cf.remoteConnect)
+	if err != nil {
+		return nil, err
+	}
+	localListenPort, err := net.ListenTCP("tcp", localTcpListenAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	l.Infof("TCP port forwarding to '%v': listening on local TCP addr: '%v'",
+		remoteTcpAddr, localTcpListenAddr)
+
+	portForwarding := &PortForwardingOutgoingTcp{
+		l:                     l,
+		tunService:            tunService,
+		cfg:                   cf,
+		localListenConnection: localListenPort,
+	}
+
+	go portForwarding.acceptOnLocalListenPort()
+
+	return portForwarding, nil
+}
+
+func (pt *PortForwardingOutgoingTcp) acceptOnLocalListenPort() error {
+	for {
+		pt.l.Debug("listening on local TCP port ...")
+		connection, err := pt.localListenConnection.AcceptTCP()
+		if err != nil {
+			fmt.Println(err)
+			return err
+		}
+
+		pt.l.Debugf("accept TCP connect from local TCP port: %v", connection.RemoteAddr())
+
+		go pt.handleClientConnection(connection)
+	}
+}
+
+func (pt *PortForwardingOutgoingTcp) handleClientConnection(localConnection *net.TCPConn) {
+	err := pt.handleClientConnection_intern(localConnection)
+	if err != nil {
+		pt.l.Debugf("Closed TCP client connection %s. Err: %v",
+			localConnection.LocalAddr().String(), err)
+	}
+}
+
+func (pt *PortForwardingOutgoingTcp) handleClientConnection_intern(localConnection net.Conn) error {
```

Comment:
Could we remove this function? Looks a bit redundant

Reply:
I like to keep it. There is a tiny but significant advantage of keeping it: The logging for the error case needs to be implemented only once instead of twice when merging the function into the caller function. I renamed the function to make this a bit more clear.